### PR TITLE
EDL security and portability enhancements

### PR DESCRIPTION
**Pull request by mikbras**
This design document proposes how to enhance EDL to address security vulnerabilities and to promote portability to TrustZone.

### CODE REVIEW

File: docs/DesignDocs/edl_security_and_portability.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+There are currently three ways that EDL *implicitly* transfers untrusted memory.
+
+1. Through ECALLs utilizing foreign structures, which may contain pointers to
+   untrusted memory.
+
+2. Through ECALLs utilizing local structures, which are not sufficiently
+   annotated for deep copy.
+
+3. Through OCALLs that return pointers.
```
Comment (anakrish) [inline docs/DesignDocs/edl_security_and_portability.md]: _user_check_ attribute also allows passing raw pointers.
Are they part of this security review?

File: docs/DesignDocs/edl_security_and_portability.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+The design is straightforward; print error diagnostics when encountering:
+
+1. Foreign structures.
+2. Under-annotated local structures.
+3. Pointer return values.
+
+Also provide an option to turn off this checking for backwards compatibility.
+
+Alternates
```
Comment (anakrish) [inline docs/DesignDocs/edl_security_and_portability.md]: Typo?

File: docs/DesignDocs/edl_security_and_portability.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+- Improve portability to TrustZone
+- Mitigate security vulnerabilities
+
+This feature is needed by TrustZone to ensure that applications developed with
+EDL will work on TrustZone platforms. Currently the use of host shared memory
+is problematic for TrustZone, which utilizes a different shared memory model.
+SGX *implicitly* shares *all* host memory with the enclave, whereas, TrustZone
+*explicitly* shares *some* host memory with the enclave. Therefore we
+distinguish two models.
```
Comment (dthaler) [inline docs/DesignDocs/edl_security_and_portability.md]: Wording issue: There may be other models for other TEEs.   For example, some likely share "none".  (For example today Windows use of TrustZone is in the "none" category, and secure elements might be in that category too.)

File: docs/DesignDocs/edl_security_and_portability.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+   code expects to write to host memory, forcing trusted memory overwrites.
+
+To prevent these kinds of errors, the stub routines must perform full
+serialization and deserialization across the trust boundary so that:
+
+- ECALL function parameters fall entirely within trusted memory.
+- OCALL function parameters fall entirely within untrusted memory.
+
+Most importantly, it should not be possible to *implicitly* pass untrusted
+memory to an ECALL function implementation. The stub routines must copy
```
Comment (dthaler) [inline docs/DesignDocs/edl_security_and_portability.md]: When you say "the stub routines", I think this should explicitly say "The trusted stub routines" for clarity.
This makes it clear you mean the _t.c part not the _u.c stubs.

File: docs/DesignDocs/edl_security_and_portability.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+3. Through OCALLs that return pointers.
+
+All three of these subtly expose untrusted memory to the enclave developer.
+
+To mitigate all three of these, the **edger8r** tool should impose three
+constraints *by default*.
+
+1. Reject foreign structures.
+2. Reject under-annotated local structures.
+3. Reject pointer return values.
```
Comment (dthaler) [inline docs/DesignDocs/edl_security_and_portability.md]: The other way to address 3 would be to propose extending EDL to allow annotating pointer return values.

File: docs/DesignDocs/edl_security_and_portability.md (hunk 1 above)
Comment (mikbras) [reply docs/DesignDocs/edl_security_and_portability.md]: Thank you Anand. I guess we could flag **user_check** uses but I was more worried about cases where we transmit host memory when the developer didn't explicitly ask for it.

File: docs/DesignDocs/edl_security_and_portability.md (hunk 2 above)
Comment (mikbras) [reply docs/DesignDocs/edl_security_and_portability.md]: Oops! Yes.

File: docs/DesignDocs/edl_security_and_portability.md (hunk 3 above)
Comment (mikbras) [reply docs/DesignDocs/edl_security_and_portability.md]: Good point. I'll update and add the "none" category.

File: docs/DesignDocs/edl_security_and_portability.md (hunk 5 above)
Comment (mikbras) [reply docs/DesignDocs/edl_security_and_portability.md]: I considered this but I thought EDL did not support function-level attribute lists.

@Andy: what do you think?

File: docs/DesignDocs/edl_security_and_portability.md (hunk 4 above)
Comment (mikbras) [reply docs/DesignDocs/edl_security_and_portability.md]: Yeah, good point. It is hard to pick a neutral terminology. :)

File: docs/DesignDocs/edl_security_and_portability.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+*explicitly* shares *some* host memory with the enclave. Therefore we
+identify three memory models.
+
+- **Implicit-all** -- implicitly shares all host memory (SGX)
+- **Explicit-some** -- explicitly shares some host memory (TrustZone)
+- **None** -- does not share any host memory (future TEEs)
+
+In TrustZone, developers may explicitly share designated memory regions. It
+is not possible to share all host memory (stack, data, heap, etc.) as it is
+in SGX.
```
Comment (HernanGatta) [inline docs/DesignDocs/edl_security_and_portability.md]: It is technically possible to map the host's memory into a TrustZone enclave. One will run into limitations imposed by OP-TEE's design, but TrustZone does indeed allow it. In other words, one could take the VA range underlying the host, translate that into a list of physical pages, and map them into secure user-mode.

File: docs/DesignDocs/edl_security_and_portability.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+This document proposes EDL enhancements that will:
+
+- Improve portability to TrustZone
+- Mitigate security vulnerabilities
+
+This feature is needed by TrustZone to ensure that applications developed with
+EDL will work on TrustZone platforms. Currently the use of host shared memory
+is problematic for TrustZone, which utilizes a different shared memory model.
+SGX *implicitly* shares *all* host memory with the enclave, whereas, TrustZone
```
Comment (HernanGatta) [inline docs/DesignDocs/edl_security_and_portability.md]: > the enclave, whereas, TrustZone

Rogue comma after "whereas".

**Comment by mikbras** [comment]
Based on Simon's comments, I wrote a new specification that is aimed at "full serialization of EDL parameters".

**Pull request closed without merging**